parquet-column/src/main/java/org/apache/parquet/schema/PrimitiveType.java

```diff
@@ -514,16 +514,41 @@ protected Type union(Type toMerge) {
     return union(toMerge, true);
   }
 
+  private void reportSchemaMergeError(Type toMerge) {
+    throw new IncompatibleSchemaModificationException("can not merge type " + toMerge + " into " + this);
+  }
+
   @Override
   protected Type union(Type toMerge, boolean strict) {
-    if (!toMerge.isPrimitive() || (strict && !primitive.equals(toMerge.asPrimitiveType().getPrimitiveTypeName()))) {
-      throw new IncompatibleSchemaModificationException("can not merge type " + toMerge + " into " + this);
+    if (!toMerge.isPrimitive()) {
+      reportSchemaMergeError(toMerge);
+    }
+
+    if (strict) {
+      // Can't merge primitive fields of different type names
+      if (!primitive.equals(toMerge.asPrimitiveType().getPrimitiveTypeName())) {
+        reportSchemaMergeError(toMerge);
+      }
+
+      // Can't merge primitive fields of different original types
+      if (getOriginalType() == null && toMerge.getOriginalType() != null ||
+          getOriginalType() != null && !getOriginalType().equals(toMerge.getOriginalType())) {
+        reportSchemaMergeError(toMerge);
```

Comment:
Since originalType is an Enum, seems we can just do like:
```
if (getOriginalType() != toMerge.getOriginalType())
  reportSchemaMergeError(toMerge);
```
which covers:
1. if both are null, the condition evals to false
2. if one is null and the other is not null, the condition evals to true
3. if both are not null, then the condition evals to false if they take different values, and evals to true otherwise.